src/udev/udev-builtin-net_id.c

```diff
@@ -600,6 +603,101 @@ static int names_platform(sd_device *dev, NetNames *names, bool test) {
+
+static int dev_devicetree_onboard(sd_device *dev, NetNames *names) {
+        char *ofnode;
+        const char *parent_syspath;
+        _cleanup_free_ char *devicetree_syspath = NULL;
+        _cleanup_free_ char *ofnode_syspath = NULL;
+        _cleanup_free_ char *ofnode_path = NULL;
+        _cleanup_closedir_ DIR *aliases = NULL;
+        sd_device *parent = NULL;
+        int r;
+        size_t ofnode_pathlen;
+        unsigned int index = 0;
+
+        if (!naming_scheme_has(NAMING_DEVICETREE_ALIASES))
+                return 0;
+
+        /* check if our direct parent has an of_node */
+        r = sd_device_get_parent(dev, &parent);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(parent, &parent_syspath);
+        if (r < 0)
+                return r;
+
+        ofnode = strjoina(parent_syspath, "/of_node");
+        r = chase_symlinks(ofnode, NULL, 0, &ofnode_syspath, NULL);
+        if (r < 0)
+                return r;
+
+        r = chase_symlinks("/proc/device-tree", NULL, 0, &devicetree_syspath, NULL);
+        if (r < 0)
+                return r;
+
+        /*
+         * Example paths:
+         * ofnode_syspath = /sys/firmware/devicetree/base/soc/ethernet@deadbeef
+         * devicetree_syspath = /sys/firmware/devicetree/base
+         * ofnode_path = soc/ethernet@deadbeef
+         */
+        r = path_make_relative(devicetree_syspath, ofnode_syspath, &ofnode_path);
+        if (r < 0)
+                return r;
+
+        if (startswith(ofnode_path, ".."))
+                /* the of_node is from another devicetree */
+                return -ENOENT;
+        ofnode_pathlen = strlen(ofnode_path);
+
+        aliases = opendir(DEVICETREE_ALIASES);
+        if (!aliases)
+                return -errno;
+
+        FOREACH_DIRENT_ALL(alias, aliases, break) {
+                const char *conflict = DEVICETREE_ALIASES "/ethernet";
+                _cleanup_free_ char *alias_path = NULL;
+                char alias_syspath[PATH_MAX];
+                size_t alias_pathlen;
+
+                if (!startswith(alias->d_name, "ethernet") ||
+                    !snprintf_ok(alias_syspath, sizeof(alias_syspath),
+                                 DEVICETREE_ALIASES "/%s", alias->d_name) ||
+                    !read_full_virtual_file(alias_syspath, &alias_path, &alias_pathlen))
+                        continue;
+
+                /* ofnode_path lacks a leading /, so we need to adjust */
+                if (ofnode_pathlen + 2 != alias_pathlen ||
+                    *alias_path != '/' ||
+                    !strneq(ofnode_path, alias_path + 1, ofnode_pathlen))
+                        continue;
+
+                /* If there's no index, we default to 0... */
+                if (streq(alias->d_name, "ethernet"))
+                        conflict = DEVICETREE_ALIASES "/ethernet0";
+                else if (sscanf(alias->d_name, "ethernet%u", &index) != 1)
+                        return -EINVAL;
```

Comment:
You can reuse the result of `startswith()` at the beginning of the loop:
```
const char *p;

p = startswith(alias->d_name, "ethernet");
if (!p)
        continue;

/* check the file contents */

if (isempty(p)) {
        index = 0;
        if (access(DEVICETREE_ALIASES "/ethernet0", F_OK) >= 0) {
                return log_device_debug_errno(dev, SYNTHETIC_ERRNO(EEXIST), "...");
else {
        r = safe_atou(p, &index);
        if (r < 0)
                return log_device_debug_errno(dev, r, "...");
}
```

Reply:
OK